drivers/misc/bcm2835_smi.c

```diff
@@ -0,0 +1,1108 @@
+{
+
+	/* SMI is inherently 32-bit, which causes surprising amounts of mess
+	   for bytes % 4 != 0. Easiest to avoid this mess altogether
+	   by handling remainder separately. */
+	int odd_bytes = n_bytes & 0x3;
+
+	n_bytes -= odd_bytes;
+	if (n_bytes > DMA_THRESHOLD_BYTES) {
+		dma_addr_t phy_addr = dma_map_single(inst->dev,
+						     buf, n_bytes,
+						     DMA_DEV_TO_MEM);
+		struct scatterlist *sgl = smi_scatterlist_from_buffer(inst,
+								      phy_addr,
+								      n_bytes);
+		if (!sgl) {
+			smi_dump_context_labelled(inst,
+			"Error: could not create scatterlist for read!");
+			return;
+		}
+		smi_dma_read_sgl(inst, sgl, 1, n_bytes);
+		dma_unmap_single(inst->dev, phy_addr, n_bytes, DMA_DEV_TO_MEM);
+	} else if (n_bytes) {
+		smi_read_fifo(inst, (uint32_t *)buf, n_bytes);
+	}
+	buf += n_bytes;
+
+	if (inst->settings.data_width == SMI_WIDTH_8BIT) {
+		while (odd_bytes--)
+			*((uint8_t *) (buf++)) = smi_read_single_word(inst);
+	} else {
+		while (odd_bytes >= 2) {
+			*(uint16_t *) buf = smi_read_single_word(inst);
+			buf += 2;
+			odd_bytes -= 2;
+		}
+		if (odd_bytes) {
+			dev_err(inst->dev,
+		"WARNING: odd number of bytes specified for wide transfer.");
+			dev_err(inst->dev,
+		"At least one byte dropped as a result.");
+			dump_stack();
+		}
+	}
+}
+EXPORT_SYMBOL(bcm2835_smi_read_buf);
+
+void bcm2835_smi_set_address(struct bcm2835_smi_instance *inst, int address)
+{
+	smi_set_address(inst, address);
+}
+EXPORT_SYMBOL(bcm2835_smi_set_address);
+
+/* Returns: -1 on fail, 0 on success.
+   Ensures only one device uses SMI at once. */
+struct bcm2835_smi_instance *bcm2835_smi_register(struct device_node *node)
+{
+	struct bcm2835_smi_instance *inst = NULL, *rv = NULL;
+	unsigned long irq_flags;
+	struct platform_device *pdev;
+
+	if (!node)
+		return NULL;
+	pdev = of_find_device_by_node(node);
+	if (!pdev)
+		return NULL;
+	inst = platform_get_drvdata(pdev);
+	if (!inst)
+		return NULL;
+
+	spin_lock_irqsave(&inst->register_lock, irq_flags);
+	if (!inst->device_registered) {
+		inst->device_registered = true;
+		rv = inst;
+	}
+	spin_unlock_irqrestore(&inst->register_lock, irq_flags);
+	return rv;
+}
+EXPORT_SYMBOL(bcm2835_smi_register);
+
+void bcm2835_smi_unregister(struct bcm2835_smi_instance *inst)
+{
+	unsigned long irq_flags;
+
+	if (inst) {
+		spin_lock_irqsave(&inst->register_lock, irq_flags);
+		inst->device_registered = false;
+		spin_unlock_irqrestore(&inst->register_lock, irq_flags);
+	}
+}
+EXPORT_SYMBOL(bcm2835_smi_unregister);
+
+/****************************************************************************
+*
+*   bcm2835_smi_probe - called when the driver is loaded.
+*
+***************************************************************************/
+
+static int bcm2835_smi_dma_setup(struct bcm2835_smi_instance *inst)
+{
+	dma_cap_mask_t mask;
+	int i, rv = 0;
+	bool alloc_failed = false;
+
+	dma_cap_zero(mask);
+	dma_cap_set(DMA_SLAVE, mask);
+	inst->dma_chan = dma_request_channel(mask, NULL, NULL);
```

Comment:
Use dma_request_slave_channel() like bcm2835-mmc does to get the dma channel from Device Tree.